### PR TITLE
Added Aggregator component

### LINKED ISSUE #15: Create component: wharf-cmd-aggregator
> Inspired by [RFC-0025: wharf-cmd provisioning](https://iver-wharf.github.io/rfcs/published/0025-wharf-cmd-provisioning)

> Depends on:
> - #10
> - #12
> - #14

Subcommand: `wharf aggregator serve`

## Requirements

- Borrow wharf-cmd-worker finding code from wharf-cmd-provisioner #10.

- Client implementations for wharf-cmd-worker's REST & gRPC APIs from #14.

- Use Kubernetes port-forwarding to connect to wharf-cmd-worker's APIs.

- Stream build results (logs, artifacts, status updates) from wharf-cmd-worker over to wharf-api.

## Out of scope (don't focus on this yet)

- Wharf-cmd-aggregator scaling. Instead assume only 1 instance.
- Support Docker as execution environment. Instead only rely on Kubernetes.
- Use of in-cluster configs. Instead only use `~/.kube/config` for local testing

### PR DESCRIPTION
**Pull request by Alexamakans**
This turned into a pretty messy PR. I think my brain has gotten stuck in the way it is right now so it feels "fine" to me. Would love some input. (Or a lot! :) ) 

---

- \[x] I've added a new note in the `CHANGELOG.md` file, according to docs:
  https://iver-wharf.github.io/#/development/changelogs/writing-changelogs

## Summary

- Added endpoint `GET /api/kill` to `workerserver`.
- Added `pkg/aggregator` and `pkg/aggregator/relayer`
  - Provides a way to stream build results from workers to Wharf API, checking for new workers on an interval.
- Added method `resultstore.Store.Freeze` that once called closes all channels (subscriptions) and blocks all write operations.
  - New subscriptions will be allowed to catch up, and are then frozen. This was needed because there was no way for the worker/server to say that it was finished, so the client (aggregator) got stuck.
- Changed to use serviceaccount for pods.

## Motivation

Closes #15

### Role
```yml
apiVersion: rbac.authorization.k8s.io/v1
kind: Role
metadata:
  name: wharf-builder
  namespace: default
rules:
- apiGroups:
  - ""
  resources:
  - pods
  verbs:
  - get
  - create
  - delete
  - watch
- apiGroups:
  - ""
  resources:
  - pods/exec
  verbs:
  - create
- apiGroups:
  - ""
  resources:
  - pods/log
  verbs:
  - get
  - list
```
### Rolebinding
```yml
apiVersion: rbac.authorization.k8s.io/v1
kind: RoleBinding
metadata:
  name: wharf-builder
  namespace: default
roleRef:
  apiGroup: rbac.authorization.k8s.io
  kind: Role
  name: wharf-builder
subjects:
- kind: ServiceAccount
  name: wharf-builder
  namespace: default
```
### Serviceaccount
```yml
apiVersion: v1
kind: ServiceAccount
metadata:
  name: wharf-builder
  namespace: default
```

### CODE REVIEW

File: pkg/provisioner/k8sprovisioner.go
Comment (Alexamakans) [inline pkg/provisioner/k8sprovisioner.go:127]: Needed this for testing, revert or change to something else?

File: pkg/provisioner/k8sprovisioner.go
Comment (Alexamakans) [inline pkg/provisioner/k8sprovisioner.go:134]: During testing, instead of downloading and installing Go to the ubuntu image I decided to use this, should I revert back to `ubuntu:20.04`?

File: pkg/provisioner/k8sprovisioner.go
Comment (jilleJr) [reply pkg/provisioner/k8sprovisioner.go:27]: If its needed for testing, I think it could stay. Doesn't matter much what it is since we don't have a real implementation is.

However `ls -alh` is obviously just for testing, while this looks like a real implementation. Need some good `// TODO: ...` comments here to show that this is not here to stay.

Same for your other "just for testing" review threads

File: pkg/provisioner/k8sprovisioner.go
Comment (Alexamakans) [reply pkg/provisioner/k8sprovisioner.go:149]: Added todo comment.

File: pkg/provisioner/k8sprovisioner.go
Comment (Alexamakans) [reply pkg/provisioner/k8sprovisioner.go:31]: Added todo comment.

**Review by jilleJr — CHANGES_REQUESTED** [review]
Pausing half-way. Will continue on monday.

File: pkg/resultstore/resultstore.go
Comment (jilleJr) [inline pkg/resultstore/resultstore.go:192]: This `Freeze` implementation is interesting. Could be good, but I think the current implementation is overly complex and error prone to race conditions.

This pubsub system isn't that easy to code around. I wonder if we could change it to rely more on file handles instead or something.

So, the flow should something like this:

- worker is done with the build
- aggregator waits until all logs and other data has been transmitted
- aggregator kills works

I'm not a fan of how this feezing logic gets a different code flow and a different freeze-check when starting the stream after the build is done.

We should do a pair-programming session for this and brainstorm some in front of a whiteboard, because I'm having issues wrapping my head around our current solution here.

File: pkg/workerapi/workerserver/server.go
Comment (jilleJr) [inline pkg/workerapi/workerserver/server.go:60]: Suggest reverting this. It needs to be passed in as a func here because the arguments to a `go` call are evaluated syncronously and not inside the new goroutine, so at line 58 it will now hault while serving the mux on the main thread, and then when it exits start `logIfErrored` inside a new goroutine, and then continue to line 59 and call `serveHTTP`...

File: pkg/workerapi/workerserver/server.go
Comment (jilleJr) [inline pkg/workerapi/workerserver/server.go:69]: Would be nice to have a timeout where it calls `s.grpc.grpc.Stop` if `GracefulStop` takes too long.

`GracefulStop` will actually wait for all streams to close themselves. If we never close the streams via our code ourselves, then `GracefulStop` will hault forever.

Maybe something like this would suffice?

```suggestion
		const timeout = 5 * time.Second
		log.Debug().WithDuration("timeout", timeout).
			Message("Attempting to shut down gracefully.")

		timer := time.AfterFunc(timeout, func() {
			log.Debug().Message("Timeout exceeded. Shutting down immediately.")
			s.grpc.grpc.Stop()
		})
		s.grpc.grpc.GracefulStop()
		timer.Stop()
```

(also need to add `"time"` to imports)

**Review by jilleJr — CHANGES_REQUESTED** [review]
Done reviewing until further changes are applied.

**Comment by jilleJr** [comment]
@Alexamakans is on vacation. I'm taking over this PR in the meantime.

File: pkg/workerapi/workerserver/server.go
Comment (jilleJr) [reply pkg/workerapi/workerserver/server.go]: Reverted.

File: pkg/aggregator/k8saggregator.go
Comment (jilleJr) [reply pkg/aggregator/k8saggregator.go:250]: Changed. Added new util type `cancelGroup` to run stuff in parallel and cancel on first fail.

File: pkg/aggregator/k8saggregator.go
Comment (jilleJr) [reply pkg/aggregator/k8saggregator.go:281]: Removed.

File: pkg/aggregator/relayer/relayer.go
Comment (jilleJr) [reply pkg/aggregator/relayer/relayer.go:10]: `pkg/aggregator/relayer` has been removed. This comment is no longer relevant.

File: pkg/aggregator/relayer/relayer.go
Comment (jilleJr) [reply pkg/aggregator/relayer/relayer.go:22]: `pkg/aggregator/relayer` has been removed. This comment is no longer relevant.

File: pkg/aggregator/relayer/sender.go
Comment (jilleJr) [reply pkg/aggregator/relayer/sender.go:12]: `pkg/aggregator/relayer` has been removed. This comment is no longer relevant.

File: pkg/workerapi/workerclient/workerclient.go
Comment (jilleJr) [reply pkg/workerapi/workerclient/workerclient.go:104]: Removed.

File: pkg/aggregator/k8saggregator.go
Comment (jilleJr) [reply pkg/aggregator/k8saggregator.go]: Created #79.

File: pkg/resultstore/logline.go
Comment (jilleJr) [inline pkg/resultstore/logline.go:71]: This freezing logic is messy. If it works, that's great. But a rewrite is in order.

Let's postpone such a rewrite for a different PR though.

My idea is that we should be able to reuse the readers for the entire reading process, and let all messages get piped through the file. In other words:

- new log line comes in from pod
- write log line to file
- (wake up any sleeping readers)
- reader continues to read and publishes all log lines via pubsub, reads until reaches EOF, and then goes to sleep
- on freeze, wake up all readers and tell them to stop on EOF instead of go to sleep.

For wake up/sleep we could use channels. Ex:

```go
func stuff() {
	sleepCh := make(chan struct{})

	// reader:
	go func() {
		for {
			// 1. read until EOF, and send all logs through pubsub
			// 2. check if should die or sleep
			<-sleepCh
		}
	}()

	// writer:
	go func() {
		for {
			// 1. wait for log line from pod
			// 2. write log line to file
			wakeUp(sleepCh)
		}
	}()
}

func wakeUp(ch chan struct{}) {
	select {
	case ch<-struct{}{}:
	default:
	}
}
```

Comment (jilleJr) [reply pkg/resultstore/logline.go]: Created #82. Resolving.

File: pkg/resultstore/resultstore.go
Comment (jilleJr) [reply pkg/resultstore/resultstore.go]: See #82

**Review by jilleJr — APPROVED** [review]
149 commits later, this is now ready.

I'm approving this because otherwise GitHub would block the merge button as I requested changes before. But as I am now maintainer of this PR, we need someone else to review this as well.

**Comment by jilleJr** [comment]
Can't request review from @Alexamakans. I'm recreating the PR instead.

**Pull request closed without merging**
